### ISSUE TITLE
(Interactive Description) Description pile-ups

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#491]
There have been a lot of improvements to the way responses (formerly knows as alerts) are handled in interactive description including the creation of the utterance queue.

Before publication with sound and sonification, we need to make sure that the timing of the descriptions is working well in Balloons and Static Electricity.

In a recent listen of these 2 versions:
- https://phet-dev.colorado.edu/html/balloons-and-static-electricity/1.5.0-dev.25/phet/balloons-and-static-electricity_en_phet.html 
- https://phet-dev.colorado.edu/html/balloons-and-static-electricity/1.5.0-dev.26/phet/balloons-and-static-electricity_en_phet.html

The descriptions were piling up during interaction.

**Comment by terracoda** [comment]
Tagging @emily-phet and @jbphet so at next meeting we can discuss if this should be a blocking issue.

**Comment by jessegreenberg** [comment]
Makes sense @terracoda. I agree, with improvements to utterance-queue since the last time this was published it seems like this is something we can do.

**Comment by jbphet** [comment]
This issue was discussed at a recent a11y meeting, and @emily-phet said that we should verify that the description behavior isn't any worse than the published version and, if it isn't, this issue shouldn't block publication.

**Comment by terracoda** [comment]
@jessegreenberg, I listened to both the published version and version 25 above.

Because of the long response strings for this sim, both versions have a tendency to become a little latent.

I think the latest version is better, though, over all. It has a better grab response, and the responses never seem to stop working, so the code is definitely better. In the published version, the responses stopped working several times.

 I think we could publish with no work on the timing of the interactive description, and we would still have a decent sim.

That said, if we have any tiny bit of combined design and dev time available before publication with sound, it would be good to optimize a little. It would be nice to have a quicker jump to the latest state on some of the longer responses, especially when rubbing against the wall, and moving off sweater.

**Comment by terracoda** [comment]
When reviewing the awesome sound design in #501, I feel that it would be time well spent to see if we can optimize the timing of the responses. 

I think the main issue is that the utterance queue is not flushing, so several responses can pile up. 

As I said in my previous comment, the over all design and functionality of the latest version on master is better than what is currently published, but some flushing of the queue of responses would be very desirable in cases of quick interaction.

**Comment by terracoda** [comment]
To maintain quality of the multi-modal experience we should really have a look to see if there may be a quick fix to the pile-ups given all the improvements that have happened to the utterance queue.

**Comment by jessegreenberg** [comment]
Discussed with @emily-phet, I will take 15 min or so to create proposal and time estimate for work that would need to be done to improve the output and we can discuss it again in meetings next week.

**Comment by jessegreenberg** [comment]
I reviewed the alerts in the sim. It looks like most alerts are already using the UtteranceQueue to collapse Utterances of the same kind. When mouse-dragging a charged balloon near the wall (~1 seconds of interaction) example output is 
![image](https://user-images.githubusercontent.com/6396244/118187887-97d7aa00-b40d-11eb-8e1b-5d72dcf2c060.png)

It takes the AT a long time, but this includes the relevant information I would expect that describes what happened without repetition.

A potential improvement could be to use `assertive` and `polite` alerts which were not supported by browsers when we released BASE but we know have support now from https://github.com/phetsims/ratio-and-proportion/issues/214. Then we would decide which alerts to make assertive. We could make the "Grabbed..." and "Released." alerts assertive so only information after those event are heard. We could make the direction change alerts assertive so that only the most recent direction is heard, and all alerts are cleared when direction changes. Or we could make the region alerts assertive so that everything before the new balloon position description is heard.

@terracoda what do you think, would this work? Which alerts should be assertive? They would be assertive in both the mouse and keyboard interactions, so this would have an impact on keyboard input too.

**Comment by emily-phet** [comment]
@terracoda @jessegreenberg Just a note on how this work relates to the publication of BASE. 
- We need to finalize the sound for BASE, so if anything related to description work could impact the sound design (e.g., the timing of when you first hear the "charges in the wall" sound and when you first hear description of behavior at the wall), we'll want to know that asap. 
- Once the sound design is finalized (which it is essentially, just awaiting final sign-off from all relevant parties) it will be waiting for a finalized Prefs menu, and any remaining description improvements prior to publication. So, if there are improvements to the description, it would be ideal to have this fit into the timeline between now and when the Prefs menu is ready for first publication.

**Comment by terracoda** [comment]
Regarding the induced charge sound. The description starts before the sonification starts.
For example when I have several more negative charges than positive charges, using the arrow keys I get 2 induced charge descriptions before I hear the sonfication.

Not sure if this is good or bad, but the description and the sound are not being triggered at the same time.

**Comment by terracoda** [comment]
Given @arouinfar's comment in a different issue and that the visual movement starts very slowly, we could consider increasing the threshold of induced charge that triggers the descriptions.

@jessegreenberg and @jbphet, could you pair on this and see if there is an easy way to align the induced charge threshold that is triggering both the sound and description for induced charge.

The fade-away seems better than the approach.

**Comment by terracoda** [comment]
@jessegreenberg, I am re-assigning back to me. I first need to review your comments on which responses need to be polite and assertive, and then we can see about aligning the sound and the description.

Apologies for the confusion.

**Comment by terracoda** [comment]
@jessegreenberg, I am reviewing the sim, and I agree it would be good to have some assertive and polite responses.

I am wondering if it makes sense to have the very short responses made into assertive responses, or the longer and more pedagogically relevant responses made assertive. 

**If short responses were assertive**
For example, some regions positions "Off Sweater", "Near Sweater", "Near Wall", "At Wall" the release response, "Released", and possibly the directional responses: "up", "left", "down" and "right", were made assertive would that keep the longer responses about position and charge from piling up. Would the longer responses jump to the latest? 

A user would have to pause to hear all the information in the latest long response, but they would get the latest info.

**If long responses were assertive**
The Grab response is very long, and if I grab the balloon and quickly move it, a lot of newer information begins to pile up.

At the moment, I can grab the balloon and deplete the entire sweater of charges before the Grab response is done talking.

If I made longer responses assertive, would I just keep hearing the beginning of each new long response, and never hear one of the shorter responses?

**Some things which I think I always want to hear are**
- Near Sweater.
- Off Sweater.
- At center of Play Area.
- At Wall (this might not be a response on its own)
- Released.
And after these events the latest info about position and charge.

**Not sure about**
- "Closer to wall"
- "Closer to sweater"
These responses are likely more helpful than the directional left, right, up, down responses, but I am not sure what would happen to other responses if I made them assertive.

@jessegreenberg, what's your availability? Maybe we can do some live pairing to try a couple of different approaches to see if there is a good way forward.

**Comment by jessegreenberg** [comment]
@terracoda  and I met today. We made the "Released",  "Grabbed" and all movement alerts assertive to reduce the pile-up. Overall it works really well and is a big improvement for VoiceOver.

While testing, we found that NVDA doesn't seem to support aria-live=assertive yet, so the changes have no impact on that platform. However, we don't expect a screen reader user to be using the mouse that frequently, and all alerts are interrupted on NVDA every key press, so the pile-up isn't as much of an issue there.

There is some information that gets cancelled now that movement alerts are assertive. For instance, when the balloon hits the sweater it is a "movement alert" so information about the charged balloon moving independently from the right side of the play area to the sweater will not be heard. We think this is OK though because the sound is conveying lots of movement information.

@terracoda  is going to test these changes on iOS VoiceOver as well, here is a dev version for that:
https://phet-dev.colorado.edu/html/balloons-and-static-electricity/1.5.0-dev.30/phet/balloons-and-static-electricity_en_phet.html

**Comment by terracoda** [comment]
@jessegreenberg, I tested again on MacOS and iOS, I think the changes we made are working well. 
I commented in #506 that, if relatively straightforward we could remove the  simple directional responses, i.e.: up, down, left, right. On mobile platform they seem to be spoken more often than when using the keyboard.

After the changes we made on Friday, the pile-ups are way less frequent. They still happen, but I think they only happen when two-balloons are active. A pile of Green balloon responses can come in after the balloons both settle down. 

**Question:**
When 1 balloon is grabbed and the other balloon is moving, are both moving responses assertive? It seems from my testing the grabbed balloon movement responses are assertive - not sure about the ungrabbed-moving balloon.

**Comment by terracoda** [comment]
@jessegreenberg, I am wondering of there is any benefit in making both assertive if they are not both assertive already. If this is not feasible or if there is no benefit, we can leave as is.

**Comment by zepumph** [comment]
RE: https://github.com/phetsims/scenery-phet/blob/8362415b66d201ba9041c81ba345dd090c25e494/js/accessibility/GrabDragInteraction.js#L285-L297

Shouldn't we make the Utterance a variable that isn't re-created each time release is called?

**Comment by zepumph** [comment]
Does this seem reasonable to you? I also added the string as a voicing response.

**Comment by jessegreenberg** [comment]
Yes, that is better thanks @zepumph.
